### ISSUE TITLE
System and user log rotation because of monotonic clock jumping backwards

### ISSUE THREAD

**Issue by lourdas** [systemd/systemd#28250]
### systemd version the issue has been seen with

253.3

### Used distribution

Gentoo

### Linux kernel version used

6.3.10-gentoo-x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-journald

### Expected behaviour you didn't see

No log rotation for system and user upon booting.

### Unexpected behaviour you saw

On every boot, system and user log rotation happens:

    Ιουλ 04 16:07:23 localhost systemd-journald[1068]: /var/log/journal/xxxxxxxxxxxxxxxxxx/system.journal: Monotonic clock jumped backwards relative to last journal entry, rotating.

and

    Ιουλ 04 16:07:44 localhost systemd-journald[1068]: /var/log/journal/xxxxxxxxxxxxxxxxxx/user-1000.journal: Monotonic clock jumped backwards relative to last journal entry, rotating.

My RTC time is set to UTC:

    $ timedatectl
                Local time: Τρι 2023-07-04 18:10:48 EEST
            Universal time: Τρι 2023-07-04 15:10:48 UTC
                  RTC time: Τρι 2023-07-04 15:10:47
                 Time zone: Europe/Athens (EEST, +0300)
    System clock synchronized: yes
               NTP service: active
           RTC in local TZ: no

### Steps to reproduce the problem

Happens always for me. I don't know what causes this.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Ιουλ 04 16:07:23 localhost kernel: rtc_cmos rtc_cmos: RTC can wake from S4
Ιουλ 04 16:07:23 localhost kernel: rtc_cmos rtc_cmos: registered as rtc0
Ιουλ 04 16:07:23 localhost kernel: rtc_cmos rtc_cmos: setting system clock to 2023-07-04T13:07:14 UTC (1688476034)
Ιουλ 04 16:07:23 localhost kernel: rtc_cmos rtc_cmos: alarms up to one month, y3k, 114 bytes nvram
```

**Comment by LW-archlinux** [comment]
Several archlinux users (including me) have the same issue, see https://bbs.archlinux.org/viewtopic.php?pid=2107990 .

On my system the first occurence of the message is at april 16 . systemd version at that date was 253.3 .

**Comment by lourdas** [comment]
All my 3 machines using systemd 253.3 and above experience the same issue. In every boot, the system and user journals are rotated because of the monotonic clock jumping backwards. My bet is that there's a bug in systemd-journald.

**Comment by OpusElectronics** [comment]
Same here, 3 machines, on Arch Linux, exact same issue on all 3 with the latest systemd 253.5.

But, things have changed with systemd 254. I just updated 2 of my 3 machines to systemd 254. The above log rotation due to the monotonic clock jumping backwards doesn't happen anymore, but now something else happens to the journals at boot, consistently.

I now get the following behavior on the two machines: upon the first reboot after updating to systemd254, the following message appears instead of the one mentioned in this thread, for both the system and user journals:
```corrupted or uncleanly shut down, renaming and replacing.```

Upon the next reboots after that, now the message becomes:
```Boot ID changed since last record, rotating.```
at every subsequent boot.

**Comment by mrc0mmand** [comment]
This seems to be fallout from https://github.com/systemd/systemd/pull/26265/commits/9204fc642acef5cfc3a411fdec3ce9a5fd9f8d37 and can be reproduced quite easily by just rebooting the machine a couple of times:

```
# journalctl --list-boots
IDX BOOT ID                          FIRST ENTRY                 LAST ENTRY                 
 -8 7e1f17ce523a4777ad1ca9d5cccb465a Sat 2023-08-05 13:19:51 UTC Sat 2023-08-05 13:20:15 UTC
 -7 2c46c912443c46198ef3a6c3019cd91b Sat 2023-08-05 13:20:20 UTC Sat 2023-08-05 13:21:54 UTC
 -6 976f37533bc8475783820f2fa3b7de33 Sat 2023-08-05 13:21:58 UTC Sat 2023-08-05 13:22:10 UTC
 -5 df8459956e724fbdb39b628a27affcda Sat 2023-08-05 13:22:15 UTC Sat 2023-08-05 13:26:03 UTC
 -4 7eaf7598f0bd41d09f5fefc7949655aa Sat 2023-08-05 13:26:08 UTC Sat 2023-08-05 13:35:41 UTC
 -3 03b4599f1d2a433eb8246854e21483d3 Sat 2023-08-05 13:35:59 UTC Sat 2023-08-05 13:36:18 UTC
 -2 f2631735d6b948f28a7c201becd1769f Sat 2023-08-05 13:36:23 UTC Sat 2023-08-05 13:38:44 UTC
 -1 3f8017673801411285f90f21231ca8c3 Sat 2023-08-05 13:38:49 UTC Sat 2023-08-05 13:43:30 UTC
  0 86d123bcd19f4a95959dae63cc8e5336 Sat 2023-08-05 13:43:35 UTC Sat 2023-08-05 13:51:35 UTC

# journalctl -b  -o short-monotonic --no-hostname -t systemd-journald
...
[    2.397512] systemd-journald[272]: Received client request to flush runtime journal.
[    2.404172] systemd-journald[272]: Fixed min_use=16.0M max_use=1.9G max_size=128.0M min_size=512.0K keep_free=1008.8M n_max_files=100
[    2.421955] systemd-journald[272]: Flushing to /var/log/journal/25d3f056400d4bab868c3ddc9ace7f60...
[    2.421979] systemd-journald[272]: Considering root directory '/run/log/journal'.
[    2.422018] systemd-journald[272]: Root directory /run/log/journal added.
[    2.422031] systemd-journald[272]: Considering directory '/run/log/journal/25d3f056400d4bab868c3ddc9ace7f60'.
[    2.422071] systemd-journald[272]: Directory /run/log/journal/25d3f056400d4bab868c3ddc9ace7f60 added.
[    2.422146] systemd-journald[272]: Journal effective settings seal=no keyed_hash=yes compress=ZSTD compress_threshold_bytes=8B
[    2.422183] systemd-journald[272]: File /run/log/journal/25d3f056400d4bab868c3ddc9ace7f60/system.journal added.
[    2.422195] systemd-journald[272]: Considering root directory '/var/log/journal'.
[    2.422198] systemd-journald[272]: Considering root directory '/var/log/journal/remote'.
[    2.425899] systemd-journald[272]: journal_file_append_entry_internal: tail_entry_boot_id: 3f8017673801411285f90f21231ca8c3; boot_id: 86d123bcd19f4a95959dae63cc8e5336
[   13.907217] systemd-journald[272]: Compressed data object 594 -> 422 using ZSTD
[   13.954477] systemd-journald[272]: Successfully sent stream file descriptor to service manager.
[   14.032997] systemd-journald[272]: Successfully sent stream file descriptor to service manager.
[   14.045858] systemd-journald[272]: Fixed min_use=141.2M max_use=1.9G max_size=128.0M min_size=512.0K keep_free=1008.8M n_max_files=100
[   14.060968] systemd-journald[272]: journal_file_append_entry_internal: tail_entry_boot_id: 3f8017673801411285f90f21231ca8c3; boot_id: 86d123bcd19f4a95959dae63cc8e5336
[   14.060976] systemd-journald[272]: Boot ID to write is different from previous boot id, refusing entry.
[   14.060991] systemd-journald[272]: Failed to write entry to /var/log/journal/25d3f056400d4bab868c3ddc9ace7f60/user-1000.journal (34 items, 935 bytes): Object is remote
[   14.060995] systemd-journald[272]: /var/log/journal/25d3f056400d4bab868c3ddc9ace7f60/user-1000.journal: Boot ID changed since last record, rotating.
[   14.060998] systemd-journald[272]: Rotating...
...
```

The
```
[    2.425899] systemd-journald[272]: journal_file_append_entry_internal: tail_entry_boot_id: 3f8017673801411285f90f21231ca8c3; boot_id: 86d123bcd19f4a95959dae63cc8e5336
```

line is from an injected debug message and shows that the `tail_entry_boot_id` contains boot ID of the previous boot.

Also, the boot indexing gets messed up as well somehow, i.e.:

```
# journalctl --list-boots
IDX BOOT ID                          FIRST ENTRY                 LAST ENTRY                 
<...snip...>
 -4 7eaf7598f0bd41d09f5fefc7949655aa Sat 2023-08-05 13:26:08 UTC Sat 2023-08-05 13:35:41 UTC

# journalctl -o export -n 1 -b -4 | grep _BOOT_ID
_BOOT_ID=df8459956e724fbdb39b628a27affcda

 -3 03b4599f1d2a433eb8246854e21483d3 Sat 2023-08-05 13:35:59 UTC Sat 2023-08-05 13:36:18 UTC

# journalctl -o export -n 1 -b -3 | grep _BOOT_ID
_BOOT_ID=7eaf7598f0bd41d09f5fefc7949655aa

 -2 f2631735d6b948f28a7c201becd1769f Sat 2023-08-05 13:36:23 UTC Sat 2023-08-05 13:38:44 UTC

# journalctl -o export -n 1 -b -2 | grep _BOOT_ID
_BOOT_ID=f2631735d6b948f28a7c201becd1769f

 -1 3f8017673801411285f90f21231ca8c3 Sat 2023-08-05 13:38:49 UTC Sat 2023-08-05 13:43:30 UTC
  0 86d123bcd19f4a95959dae63cc8e5336 Sat 2023-08-05 13:43:35 UTC Sat 2023-08-05 13:51:35 UTC
```

(on my other testing machine the indexing was messed up even more)

/cc @poettering

**Comment by vanillaSprinkles** [comment]
similar issues on systemd 254.1; some fun details:
- archlinux, linux-kernel 6.4.12
- emmc os storage medium, btrfs

- `journalctl -b` was not showing kernel messages that show up in dmesg (particularly early boot messages up to 90 - 140 seconds in)
- `journalctl --field _TRANSPORT` not showing `driver` nor `audit` 
- use the following unit override adds `audit` back in only `/etc/systemd/system/systemd-journald.service.d/override.conf`
  ```
 [Service]
Sockets=
Sockets=systemd-journald.socket systemd-journald-dev-log.socket systemd-journald-audit.socket
```


downgrade to `systemd 253 (253.4-1-arch)`
attempted multiple reboots with and without the above mentioned systemd unit override
- kernel messages seem to be logged intermittently to journald
  times correspond to `Monotonic clock jumped backwards relative to last journal entry, rotating.`
- sometimes the journal claims to be corrupted (this is even after a `--vacuum-time=1s`

**Comment by yuwata** [comment]
Should be already fixed by addcecf6f4686811b63d48d51398d887c2c4bb68.

**Comment by yuwata** [comment]
I forget why I stroke out the previous comment, but this should be already fixed by addcecf6f4686811b63d48d51398d887c2c4bb68. Closing.

Please request to reopen this page if you still observe the issue.

**Comment by mrc0mmand** [comment]
FTR I tried to reproduce the original issue while reproducing #30092 and it indeed seems to be gone.
